### PR TITLE
OCLOMRS-469: When one clicks Cancel (Edit Concept page) show message a “Are you sure you want to cancel without saving ?

### PR DESCRIPTION
**Pull request by alexkayabula**
# JIRA TICKET NAME:
[When one clicks Cancel (Edit Concept page) show message a “Are you sure you want to cancel without saving ?”](https://issues.openmrs.org/browse/OCLOMRS-469)

# Summary:
On the Edit Concept page, when one clicks "Cancel", there should be a message saying, "Are you sure you want to cancel without saving?" with options to proceed or not.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -30,7 +30,7 @@
     "uuid": "^3.3.2"
   },
   "scripts": {
-    "start": "chmod +x ./init.sh && ./init.sh && cp env-config.js ./public/ && npm run build:css && react-scripts start",
```
Comment (dkayiwa) [inline package.json]: How is this related to the ticket?

File: src/tests/dictionaryConcepts/container/EditConcept.test.jsx [hunk 2]
```diff
@@ -61,9 +62,48 @@ describe('Test suite for dictionary concepts components', () => {
...
+    const instance = wrapper.find('EditConcept').instance();
+    instance.showModal();
+    expect(instance.state.show).toEqual(true);
+  });
+
+  it('should redirect on cancel', () => {
+    const wrapper = mount(<Router>
+      <EditConcept {...props} isChange={false} />
+    </Router>);
+    const instance = wrapper.find('EditConcept').instance();
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: Where are you asserting for this test?

File: package.json (hunk 1 above)
Comment (alexkayabula) [reply package.json]: It is not related in any way but the reason why I changed that was due to the fact the application would not running with that included.

Comment (dkayiwa) [reply package.json]: That requires a separate ticket and pull request.

Comment (alexkayabula) [reply package.json]: I am going to let the sprint lead know and we create the ticket.

File: src/tests/dictionaryConcepts/container/EditConcept.test.jsx (hunk 2 above)
Comment (alexkayabula) [reply src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: I have adjusted the test.

Comment (alexkayabula) [reply src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: I fixed the issue, I am asserting the value of "isChange" to be false when the "selectConfirm" method is called.

File: package.json (hunk 1 above)
Comment (dkayiwa) [reply package.json]: If it was addressed in that ticket, then what is this change still doing in your pull request?

Comment (alexkayabula) [reply package.json]: I have rebased with master and fixed that issue.

File: src/tests/dictionaryConcepts/container/EditConcept.test.jsx [hunk 3]
```diff
@@ -61,9 +62,42 @@ describe('Test suite for dictionary concepts components', () => {
     fetchAllConceptSources: jest.fn(),
     allSources: [mockSource],
     addSelectedAnswers: jest.fn(),
+    selectConfirm: jest.fn(),
     selectedAnswers: [],
     createNewAnswerRow: jest.fn(),
   };
+
+  it('should call selectConfirm', () => {
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: When should select confirm be called?

File: src/tests/dictionaryConcepts/container/EditConcept.test.jsx (hunk 2 above)
Comment (dkayiwa) [reply src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: Where are you testing for the redirect?

File: src/tests/dictionaryConcepts/container/EditConcept.test.jsx (hunk 3 above)
Comment (alexkayabula) [reply src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: "select confirm" is called when confirming the cancellation on the pop-up cancel modal.

Comment (dkayiwa) [reply src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: Would the maintainer of this code always be looking out for you to get that answer?

File: src/tests/dictionaryConcepts/container/EditConcept.test.jsx (hunk 2 above)
Comment (alexkayabula) [reply src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: I am testing for the redirect at the point where the instance of the "selectConfirm" method is called, given the value of "isChange" is false.

File: src/tests/dictionaryConcepts/container/EditConcept.test.jsx (hunk 3 above)
Comment (alexkayabula) [reply src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: No, I have made the title of the test more appropriate. Thank you for your observation.

File: src/tests/dictionaryConcepts/container/EditConcept.test.jsx (hunk 2 above)
Comment (dkayiwa) [reply src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: How does that prove it is a redirect?

Comment (alexkayabula) [reply src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: I have adjusted the test.

File: src/tests/dictionaryConcepts/container/EditConcept.test.jsx [hunk 4]
```diff
@@ -61,9 +62,45 @@ describe('Test suite for dictionary concepts components', () => {
...
+      <EditConcept {...props} history={history} />
+    </Router>);
+    const instance = wrapper.find('EditConcept').instance();
+    wrapper.find('.btn-danger').at(2).simulate('click');
+    instance.selectConfirm();
+    expect(instance.state.isChange).toEqual(false);
+    expect(history.push).toHaveBeenCalledWith('/concepts/users/admin/CCL/CASE CLINIC/en');
+  });
+
+  it('should hide the cancel modal', () => {
```
Comment (dkayiwa) [inline src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: When should the cancel model be hidden?

Comment (alexkayabula) [reply src/tests/dictionaryConcepts/container/EditConcept.test.jsx]: I have adjusted the test title.

**Comment by dkayiwa** [comment]
Why does it say that you have un saved changes when there is not a single change made?

**Comment by alexkayabula** [comment]
> Why does it say that you have un saved changes when there is not a single change made?

Let me double check everything and I get back to you.

File: src/components/dictionaryConcepts/components/CreateConceptForm.jsx [hunk 5]
```diff
@@ -298,6 +297,9 @@ CreateConceptForm.defaultProps = {
   isEditConcept: false,
   handleAsyncSelectChange: () => {},
   handleAnswerChange: () => {},
+  // eslint-disable-next-line react/default-props-match-prop-types
```
Comment (paulvazbad) [inline src/components/dictionaryConcepts/components/CreateConceptForm.jsx]: Is this line necessary for the commit or just a local setting to your linter?

File: src/components/dictionaryConcepts/containers/EditConcept.jsx [hunk 6]
```diff
@@ -348,6 +353,20 @@ export class EditConcept extends Component {
     this.hideGeneralModal();
   }
 
+  selectConfirm = () => this.props.history.push(localStorage.getItem('dictionaryPathName'));
+
```
Comment (paulvazbad) [inline src/components/dictionaryConcepts/containers/EditConcept.jsx]: A null check should be performed on the localStorage.getItem(...)

Comment (alexkayabula) [reply src/components/dictionaryConcepts/containers/EditConcept.jsx]: @paulvazbad, what is the essence of performing a null check on the localStorage.getItem

File: src/components/dictionaryConcepts/components/CreateConceptForm.jsx (hunk 5 above)
Comment (alexkayabula) [reply src/components/dictionaryConcepts/components/CreateConceptForm.jsx]: Fixed

File: src/components/dictionaryConcepts/containers/EditConcept.jsx [hunk 7]
```diff
@@ -87,6 +86,8 @@ export class EditConcept extends Component {
       openGeneralModal: false,
       url: '',
       mapp: '',
+      isChange: false,
```
Comment (dkayiwa) [inline src/components/dictionaryConcepts/containers/EditConcept.jsx]: Do you still need these flags?

Comment (alexkayabula) [reply src/components/dictionaryConcepts/containers/EditConcept.jsx]: I guess there is no need, I have refactored and removed them.

File: src/components/dictionaryConcepts/containers/EditConcept.jsx (hunk 6 above)
Comment (alexkayabula) [reply src/components/dictionaryConcepts/containers/EditConcept.jsx]: I have refactored and fixed the issue.

**Comment by alexkayabula** [comment]
> > Why does it say that you have un saved changes when there is not a single change made?
> 
> Let me double check everything and I get back to you.

I have refactored accordingly.
